# Shark Attack Project

## Objetivo do Projeto:
   - Limpar os dados da tabela "Global Shark Attack"
   - Fazer uma análise descritiva dos ataques de tubarões no mundo.
   - Levantar informações para uma ação de comunicação que visa reduzir a incidência 
    dos ataques.

# Baixando e analisando o arquivo:
--------------------------------------------------------------------------------------------
## Baixar e dar uma olhada no arquivo do Kaggle sobre ataques de tubarão.

In [16]:
import pandas as pd
import numpy as np

In [17]:
atck_full= pd.read_csv('Datas/attacks.csv')
atck_full.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson�McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [18]:
atck_full

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson�McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

In [19]:
atck_full.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [20]:
atck_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

# Limpando os dados:
---------------------------------------------------------------------------------------------

## Ver tudo o que tem no arquivo e começar a limpeza dos dados.

In [21]:
for column in atck_full:
    if atck_full[column].dtypes.type == np.object_:
        print(atck_full[column].describe())

count     8702
unique    6287
top          0
freq      2400
Name: Case Number, dtype: object
count     6302
unique    5433
top       1957
freq        11
Name: Date, dtype: object
count           6298
unique             8
top       Unprovoked
freq            4595
Name: Type, dtype: object
count     6252
unique     212
top        USA
freq      2229
Name: Country, dtype: object
count        5847
unique        825
top       Florida
freq         1037
Name: Area, dtype: object
count                                 5762
unique                                4108
top       New Smyrna Beach, Volusia County
freq                                   163
Name: Location, dtype: object
count        5758
unique       1532
top       Surfing
freq          971
Name: Activity, dtype: object
count     6092
unique    5230
top       male
freq       550
Name: Name, dtype: object
count     5737
unique       6
top          M
freq      5094
Name: Sex , dtype: object
count     3471
unique     157
top         17
fre

### Criando um DataFrame com as colunas que tem dados descritivos de interesse.

In [22]:
atck = (atck_full['Year'], atck_full['Type'], atck_full['Country'], atck_full['Area'],
        atck_full['Activity'],atck_full['Sex '], atck_full['Age'], atck_full['Injury'],
        atck_full['Fatal (Y/N)'], atck_full['Species '])

In [23]:
atck = pd.DataFrame(atck)
atck = atck.transpose()
atck.head()

Year        Type    Country             Area     Activity Sex   Age  \
0  2018.0     Boating        USA       California     Paddling    F   57   
1  2018.0  Unprovoked        USA          Georgia     Standing    F   11   
2  2018.0     Invalid        USA           Hawaii      Surfing    M   48   
3  2018.0  Unprovoked  AUSTRALIA  New South Wales      Surfing    M  NaN   
4  2018.0    Provoked     MEXICO           Colima  Free diving    M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

          Species   
0      White shark  
1              NaN  
2              NaN  
3        2 m shark  
4  Tiger shark, 3m

In [24]:
atck

Year        Type    Country             Area     Activity Sex   Age  \
0      2018.0     Boating        USA       California     Paddling    F   57   
1      2018.0  Unprovoked        USA          Georgia     Standing    F   11   
2      2018.0     Invalid        USA           Hawaii      Surfing    M   48   
3      2018.0  Unprovoked  AUSTRALIA  New South Wales      Surfing    M  NaN   
4      2018.0    Provoked     MEXICO           Colima  Free diving    M  NaN   
...       ...         ...        ...              ...          ...  ...  ...   
25718     NaN         NaN        NaN              NaN          NaN  NaN  NaN   
25719     NaN         NaN        NaN              NaN          NaN  NaN  NaN   
25720     NaN         NaN        NaN              NaN          NaN  NaN  NaN   
25721     NaN         NaN        NaN              NaN          NaN  NaN  NaN   
25722     NaN         NaN        NaN              NaN          NaN  NaN  NaN   

                                                  Injury Fatal (Y/N)  \
0      No injury to occupant, outrigger canoe and pad...           N   
1                             Minor injury to left thigh           N   
2           Injury to left lower leg from surfboard skeg           N   
3                              Minor injury to lower leg           N   
4      Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                  ...         ...   
25718                                                NaN         NaN   
25719                                                NaN         NaN   
25720                                                NaN         NaN   
25721                                                NaN         NaN   
25722                                                NaN         NaN   

              Species   
0          White shark  
1                  NaN  
2                  NaN  
3            2 m shark  
4      Tiger shark, 3m  
...                ...  
25718              NaN  
25719              NaN  
25720              NaN  
25721              NaN  
25722              NaN  

[25723 rows x 10 columns]

In [25]:
print(atck['Year'].describe())
print('')
print(atck['Year'].value_counts())

count     6300.0
unique     249.0
top       2015.0
freq       143.0
Name: Year, dtype: float64

2015.0    143
2017.0    136
2016.0    130
2011.0    128
2014.0    127
         ... 
1807.0      1
1805.0      1
1804.0      1
1802.0      1
1791.0      1
Name: Year, Length: 249, dtype: int64


In [26]:
# Transformar a coluna 'Year', de object para float
atck['Year'] = atck['Year'].astype(float, errors = 'raise')

# A primeira linha está classificado como 'No Shark involvement', vou deletar. 
# IMPORTANTE: depois ver se tem mais 'No Shark involvement' na coluna 'Injury'!!!
atck = atck.drop(0, axis = 0)
atck = atck.reset_index(drop = True)

atck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25722 entries, 0 to 25721
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6299 non-null   float64
 1   Type         6297 non-null   object 
 2   Country      6251 non-null   object 
 3   Area         5846 non-null   object 
 4   Activity     5757 non-null   object 
 5   Sex          5736 non-null   object 
 6   Age          3470 non-null   object 
 7   Injury       6273 non-null   object 
 8   Fatal (Y/N)  5762 non-null   object 
 9   Species      3463 non-null   object 
dtypes: float64(1), object(9)
memory usage: 2.0+ MB


### Criar um DataFrame: data mais recente para a mais antiga, ordenado pela idade de forma crescente.

In [27]:
# Organizar a tabela de acordo com o ano em que aconteceu o ataque, em ordem decrescente -
# e pela idade em ordem crescente
# Resetar a indexação das linhas
atck = atck.sort_values(['Year', 'Age'], ascending = [False, True])
atck = atck.reset_index(drop = True)

# Criar uma coluna com a indexação do DataFrame para facilitar a encontrar os erros depois
atck['index'] = atck.index

## Limpando a Coluna 'Age'
--------------------------------------------------------
### Tirar os elementos da coluna 'Age' que apresentam dúvida de idade. Ex: 18 or 20.
-------------------------------------------------
Vou utilizar a menor idade. Alteração não significativa para a estatística final.

In [29]:
# Encontrar todos os elementos da atck['Age'] com idade 'x or y'
import regex as re

padrão = '[ 0-9]+[ or]+[ 0-9]+'
lista_or = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_or.append(i)
    except:
        None
print(lista_or)
print(len(lista_or))

['18 or 20', '12 or 13', '8 or 10', '30 or 36', '33 or 37', '21 or 26', '25 or 28', '13 or 18', '7 or 8', '8 or 10', '9 or 10', '10 or 12', '31 or 33', '7 or 8', '13 or 14']
15


In [30]:
# Criar uma lista com o index de cada linha de casos múltiplos
lista_index = list()
atck_copy = atck[:]

for i in lista_or:
    g = atck_copy[atck_copy['Age'] == i]
    h = g['index']
    if len(h) == 1:
        lista_index.append(int(h.values))
        atck_copy.drop(int(h.values))
    elif len(h) > 1:
        for i in h:
            ab = h[i]
            lista_index.append(ab)
                
# Apareceram na lista valores dobrados. Vou criar um set para corrigir.

set_index = set(lista_index)
print(set_index)
print(len(set_index))


{1314, 1413, 4037, 4616, 1931, 4812, 6061, 5291, 1776, 3891, 1845, 4471, 5464, 6174, 3166}
15


In [31]:
# Utilizar o set_index para alterar a idade das células dúbias.
for i in set_index:
    atck['Age'].iloc[i] = atck['Age'].iloc[i][0] + atck['Age'].iloc[i][1]

# Células dúbias da coluna 'Age' resolvidas!


C:\Users\guoco\anaconda3\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [32]:
# Confirmando que deu certo...
atck.iloc[4471]

Year                                              1950.0
Type                                          Unprovoked
Country                                        INDONESIA
Area                                     Jakarta Harbour
Activity                                    Bending over
Sex                                                    M
Age                                                   7 
Injury                                             FATAL
Fatal (Y/N)                                            Y
Species        Tiger shark, 3.7 m to 4.3 m  [12' to 14']
index                                               4471
Name: 4471, dtype: object

### Dividindo informações múltiplas em uma mesma linha
----------------------------------------------------------------------------------
Encontrei algumas linhas com informações multiplas, ou seja, mais de um acidente registrado em uma única informação.



### Começar identificando todas as linhas com informações múltiplas. 
 - Encontrar as linhas com informações múltiplas
 - Registrar a indexação delas
 - Separar a informação múltipla em informações únicas

In [33]:
# Encontrar as linhas que tem 2 vítimas registradas

import regex as re

padrão = '[ 0-9]+[, &]+[ 0-9]+'
lista_mult2 = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_mult2.append(i)
    except:
        None
print(lista_mult2)
print(len(lista_mult2))

['28 & 26', '46 & 34', '36 & 26', '23 & 20', '7      &    31', '32 & 30', '9 & 12', '23 & 26', '33 & 37', '30 & 32', '17 & 35', '33 & 26', '34 & 19', '50 & 30', '17 & 16', '36 & 23']
16


In [34]:
# Econtrar a indexação com as informações das linhas com informações múltiplas com 2 vítimas
# Vou precisar dessa informação para criar as linhas com informações únicas correspondentes
lista_index2 = list()

for elemento in lista_mult2:
    a = atck.loc[atck['Age'] == elemento]
    lista_index2.append(int(a['index']))

print(lista_index2)
print(len(lista_index2))

[623, 1478, 1758, 2000, 2432, 2480, 3475, 3731, 3896, 3966, 4045, 4071, 4072, 4076, 4562, 4977]
16


In [35]:
# Transformar cada string da lista em uma nova lista com 2 elementos:
# Cada elemento significa a idade individual de cada vítima de ataque
padrão = '[0-9]+'
lista_ind2 = list()

for i in lista_mult2:
    a = re.findall(padrão, i)
    lista_ind2.append(a)

print(lista_ind2)
print(len(lista_ind2))

[['28', '26'], ['46', '34'], ['36', '26'], ['23', '20'], ['7', '31'], ['32', '30'], ['9', '12'], ['23', '26'], ['33', '37'], ['30', '32'], ['17', '35'], ['33', '26'], ['34', '19'], ['50', '30'], ['17', '16'], ['36', '23']]
16


In [36]:
# Transformar os str em int para 2 casos
count = 0

while count < 16:
    for i in lista_ind2:
        lista_ind2[count][0] = int(lista_ind2[count][0])
        lista_ind2[count][1] = int(lista_ind2[count][1])
        count += 1
        
print(lista_ind2)
print(len(lista_ind2))

[[28, 26], [46, 34], [36, 26], [23, 20], [7, 31], [32, 30], [9, 12], [23, 26], [33, 37], [30, 32], [17, 35], [33, 26], [34, 19], [50, 30], [17, 16], [36, 23]]
16


In [37]:
# Encontrar as linhas que tem 3 vítimas registradas
padrão = '[ 0-9]+[, &]+[ 0-9]+[, &]+[ 0-9]+'
lista_mult3 = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_mult3.append(i)
    except:
        None
print(lista_mult3)
lista_mult3.pop()
print(lista_mult3)

['28, 23 & 30', '7      &    31']
['28, 23 & 30']


In [38]:
# Econtrar a indexação com as informações das linhas com informações múltiplas com 3 vítimas
# Vou precisar dessa informação para criar as linhas com informações únicas correspondentes
lista_index3 = list()

for elemento in lista_mult3:
    a = atck.loc[atck['Age'] == elemento]
    lista_index3.append(int(a['index']))

print(lista_index3)


[1456]


In [39]:
# Transformar cada string da lista em uma nova lista com 3 elementos:
# Cada elemento significa a idade individual de cada vítima de ataque
padrão = '[0-9][0-9]'
lista_ind3 = list()

for i in lista_mult3:
    a = re.findall(padrão, i)
    lista_ind3.append(a)

print(lista_ind3)

[['28', '23', '30']]


In [40]:
# Transformar os str em int para 3 casos

lista_ind3[0][0] = int(lista_ind3[0][0])
lista_ind3[0][1] = int(lista_ind3[0][1])
lista_ind3[0][2] = int(lista_ind3[0][2])  
        
print(lista_ind3)


[[28, 23, 30]]


In [41]:
# Encontrar as linhas que tem 4 vítimas registradas
padrão = '[ 0-9]+[, &]+[ 0-9]+[, &]+[ 0-9]+[, &]+[ 0-9]+'
lista_mult4 = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_mult4.append(i)
    except:
        None
print(lista_mult4)
lista_mult4.pop(0)
print(lista_mult4)

['7      &    31', '21, 34,24 & 35']
['21, 34,24 & 35']


In [42]:
# Econtrar a indexação com as informações das linhas com informações múltiplas com 4 vítimas
# Vou precisar dessa informação para criar as linhas com informações únicas correspondentes
lista_index4 = list()

for elemento in lista_mult4:
    a = atck.loc[atck['Age'] == elemento]
    lista_index4.append(int(a['index']))

print(lista_index4)

[3951]


In [43]:
# Transformar cada string da lista em uma nova lista com 4 elementos:
# Cada elemento significa a idade individual de cada vítima de ataque
padrão = '[0-9][0-9]'
lista_ind4 = list()

for i in lista_mult4:
    a = re.findall(padrão, i)
    lista_ind4.append(a)

print(lista_ind4)

[['21', '34', '24', '35']]


In [44]:
# Transformar os str em int para 4 casos

lista_ind4[0][0] = int(lista_ind4[0][0])
lista_ind4[0][1] = int(lista_ind4[0][1])
lista_ind4[0][2] = int(lista_ind4[0][2])  
lista_ind4[0][3] = int(lista_ind4[0][3])  

print(lista_ind4)

[[21, 34, 24, 35]]


### Criando listas com informações únicas a partir das linhas com informações múltiplas

In [45]:
# Criar uma linha com as informações da linha múltipla
# Substituir o valor de 'Age', pelas informações únicas.
linha_623 = list()
linha_623.append(atck.iloc[623, 0])
linha_623.append(atck.iloc[623, 1])
linha_623.append(atck.iloc[623, 2])
linha_623.append(atck.iloc[623, 3])
linha_623.append(atck.iloc[623, 4])
linha_623.append(atck.iloc[623, 5])
linha_623.append(atck.iloc[623, 6])
linha_623.append(atck.iloc[623, 7])
linha_623.append(atck.iloc[623, 8])
linha_623.append(atck.iloc[623, 9])

print(linha_623)

linha_623_1 = linha_623[:]
linha_623_1[6] = 28

linha_623_2 = linha_623[:]
linha_623_2[6] = 26



print(linha_623_1)

print(linha_623_2)


# Deu certo! Agora precisa criar funções para fazer isso em grande escala!
# Depois precisa adicionar essas linhas no DataFrame e retirar as multiplas!

[2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', '28 & 26', 'Minor leg injuries', 'N', '"small sharks"']
[2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', 28, 'Minor leg injuries', 'N', '"small sharks"']
[2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', 26, 'Minor leg injuries', 'N', '"small sharks"']


In [46]:

def info_mult(linha):
    '''linha = index do DataFrame. Cria uma lista com todas as informações daquele DataFrame.
    Mais fácil trabalhar com lista do que com pd.series.'''
    
    infos = list()
    global atck
    infos.append(atck.iloc[linha, 0])
    infos.append(atck.iloc[linha, 1])
    infos.append(atck.iloc[linha, 2])
    infos.append(atck.iloc[linha, 3])
    infos.append(atck.iloc[linha, 4])
    infos.append(atck.iloc[linha, 5])
    infos.append(atck.iloc[linha, 6])
    infos.append(atck.iloc[linha, 7])
    infos.append(atck.iloc[linha, 8])
    infos.append(atck.iloc[linha, 9])    
    
    return infos

In [47]:
# Criar uma lista de listas com todas as linhas que estão múltiplas com 2 informações
lista2_junt = list()

for i in lista_index2:
    a = info_mult(i)
    lista2_junt.append(a)
    
print(lista2_junt)
print(len(lista2_junt))

[[2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', '28 & 26', 'Minor leg injuries', 'N', '"small sharks"'], [2006.0, 'Boating', 'ATLANTIC OCEAN', '300 miles from Antigua', 'Competing in the Woodvale Atlantic Rowing Race', 'M', '46 & 34', 'No injury to occupants; shark rammed boat repeatedly', 'N', "12' shark"], [2003.0, 'Boat', 'SOUTH AFRICA', 'Western Cape Province', 'Fishing ', nan, '36 & 26', 'No injury to occupants, shark bit boat', 'N', '2 m cow shark'], [2000.0, 'Invalid', 'USA', 'Florida', 'Attempting to illegally enter the USA', 'M', '23 & 20', 'Shark involvement probably post-mortem', nan, 'Shark involvement prior to death was not confirmed'], [1994.0, 'Sea Disaster', 'USA', 'Florida', 'Adrift on refugee raft', 'M', '7      &    31', 'FATAL', 'Y', nan], [1993.0, 'Boating', 'CARIBBEAN SEA', nan, '   ', 'M', '32 & 30', 'No injury to occupants. Sharks, attracted to offal from slaughtered dolphin, attacke

In [48]:
# Criar a lista de listas com todas as linhas com 2 informações separadas 
# no DataFrame atck que estão prontas para serem inseridas

lista_age_final = list()
count = 0

for i in lista2_junt:
    b = i[:]
    a = i[:]
    b[6] = lista_ind2[count][0]
    lista_age_final.append(b)
    a[6] = lista_ind2[count][1]
    lista_age_final.append(a)
    count += 1

print(lista_age_final)
print(len(lista_age_final))

[[2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', 28, 'Minor leg injuries', 'N', '"small sharks"'], [2013.0, 'Unprovoked', 'PHILIPPINES', 'Palawan', 'Swimming to shore with floatioon devices after boat engine conked out', 'M', 26, 'Minor leg injuries', 'N', '"small sharks"'], [2006.0, 'Boating', 'ATLANTIC OCEAN', '300 miles from Antigua', 'Competing in the Woodvale Atlantic Rowing Race', 'M', 46, 'No injury to occupants; shark rammed boat repeatedly', 'N', "12' shark"], [2006.0, 'Boating', 'ATLANTIC OCEAN', '300 miles from Antigua', 'Competing in the Woodvale Atlantic Rowing Race', 'M', 34, 'No injury to occupants; shark rammed boat repeatedly', 'N', "12' shark"], [2003.0, 'Boat', 'SOUTH AFRICA', 'Western Cape Province', 'Fishing ', nan, 36, 'No injury to occupants, shark bit boat', 'N', '2 m cow shark'], [2003.0, 'Boat', 'SOUTH AFRICA', 'Western Cape Province', 'Fishing ', nan, 26, 'No injury to occupants, s

In [49]:
# Criar uma lista de listas com a linha múltipla de 3 informações
print(lista_index3)

lista3 = info_mult(1456)

print(lista3)

[1456]
[2006.0, 'Boating', 'ATLANTIC OCEAN', '800 miles from land', 'Competing in the Woodvale Atlantic Rowing Race', 'F', '28, 23 & 30', 'No injury to occupants; a shark, accidentally caught in a line, threatened to rip out the transom', 'N', nan]


In [50]:
# Criar uma lista com as informações isoladas da lista3
# Acrescentar essas listas a lista_age_final
print(lista3[6])
x = lista3[:]
y = lista3[:]
z = lista3[:]

x[6] = 28
y[6] = 23
z[6] = 30

lista_age_final.append(x)
lista_age_final.append(y)
lista_age_final.append(z)

print(len(lista_age_final))

28, 23 & 30
35


In [51]:
# Criar uma lista de listas com a linha múltipla de 4 informações
print(lista_index4)

lista4 = info_mult(3951)

print(lista4)

[3951]
[1960.0, 'Invalid', 'BERMUDA', '33N, 68W', 'Floating on a raft', 'M', '21, 34,24 & 35', 'No injury, 5 sharks bumped raft', nan, 'Questionable incident']


In [52]:
# Criar uma lista com as informações isoladas da lista4
# Acrescentar essas listas a lista_age_final
print(lista4[6])
p = lista4[:]
q = lista4[:]
r = lista4[:]
s = lista4[:]

p[6] = 21
q[6] = 34
r[6] = 24
s[6] = 35

lista_age_final.append(p)
lista_age_final.append(q)
lista_age_final.append(r)
lista_age_final.append(s)

print(len(lista_age_final))


21, 34,24 & 35
39


In [53]:
# Informações múltiplas de 5 casos em uma única linha = index 3986
# Idades = 37, 67, 35, 27, ? & 27 - Desconsiderar o que não tem informação

lista5 = info_mult(3971)
print(lista5)
print(lista5[6])

ab = lista5[:]
ac = lista5[:]
ad = lista5[:]
ae = lista5[:]
af = lista5[:]

ab[6] = 37
ac[6] = 67
ad[6] = 35
ae[6] = 27
af[6] = 27

lista_age_final.append(ab)
lista_age_final.append(ac)
lista_age_final.append(ad)
lista_age_final.append(ae)
lista_age_final.append(af)

print(len(lista_age_final))

[1960.0, 'Invalid', 'USA', 'California', 'Testing classified underwater electronic gear for Raytheon Corporation, vessel torn apart by explosion', 'M', '37, 67, 35, 27,  ? & 27', 'Legs & arms bitten, coroner unable to determine if injuries occurred before death.', nan, 'Shark involvement prior to death was not confirmed']
37, 67, 35, 27,  ? & 27
44


### Removendo as linhas dúbias e múltiplas do DataFrame

In [54]:
print(lista_index2)

[623, 1478, 1758, 2000, 2432, 2480, 3475, 3731, 3896, 3966, 4045, 4071, 4072, 4076, 4562, 4977]


In [55]:
# Tirando a coluna de indexação para depois poder agrupar os DataFrames
atck = atck.drop(columns=['index'])

In [56]:
# Deletando as linhas com 2 casos registrados em uma única linha
for i in lista_index2:
    atck = atck.drop(i)

In [57]:
# testando se a idade dobrada foi apagada. tudo certo...
atck.iloc[3971]

Year                                                      1960.0
Type                                                     Invalid
Country                                                      USA
Area                                                     Florida
Activity                        Fell overboard, prop slashed arm
Sex                                                            F
Age                                                           51
Injury         FATAL, probable drowning & post mortem scavenging
Fatal (Y/N)                                                  NaN
Species        Shark involvement prior to death was not confi...
Name: 3981, dtype: object

In [58]:
# Deletando as linhas com 3 e 4 casos registrados em uma única linha
# lista_index3 = 1456
# lista_index4 = 3951
# lista_index5 = 

atck = atck.drop(1456)
atck = atck.drop(3951)
atck = atck.drop(3971)

In [59]:
# Testando se o drop funcionou. ok...
atck.iloc[25702]

Year           NaN
Type           NaN
Country        NaN
Area           NaN
Activity       NaN
Sex            NaN
Age            NaN
Injury         NaN
Fatal (Y/N)    NaN
Species        NaN
Name: 25721, dtype: object

In [60]:
atck

Year        Type    Country               Area         Activity Sex   \
0      2018.0  Unprovoked        USA     South Carolina  Boogie boarding    M   
1      2018.0  Unprovoked        USA     South Carolina         Swimming    M   
2      2018.0  Unprovoked  AUSTRALIA  Western Australia           Wading    M   
3      2018.0  Unprovoked        USA            Georgia         Standing    F   
4      2018.0  Unprovoked        USA            Florida         Standing    M   
...       ...         ...        ...                ...              ...  ...   
25717     NaN         NaN        NaN                NaN              NaN  NaN   
25718     NaN         NaN        NaN                NaN              NaN  NaN   
25719     NaN         NaN        NaN                NaN              NaN  NaN   
25720     NaN         NaN        NaN                NaN              NaN  NaN   
25721     NaN         NaN        NaN                NaN              NaN  NaN   

       Age                                Injury Fatal (Y/N)         Species   
0       10  Injuries to lower right leg and foot           N              NaN  
1       10          Severe bite to right forearm           N              NaN  
2       10                Minor injuries to legs           N  Wobbegong shark  
3       11            Minor injury to left thigh           N              NaN  
4       12                  Minor injury to foot           N              NaN  
...    ...                                   ...         ...              ...  
25717  NaN                                   NaN         NaN              NaN  
25718  NaN                                   NaN         NaN              NaN  
25719  NaN                                   NaN         NaN              NaN  
25720  NaN                                   NaN         NaN              NaN  
25721  NaN                                   NaN         NaN              NaN  

[25703 rows x 10 columns]

### Adiconando as informações única no DataFrame principal (atck)

In [61]:
# Criando um DataFrame com as mesma colunas que o atck para concatenar.
age_final = pd.DataFrame(lista_age_final)
age_final.columns = atck.columns
age_final

Year          Type             Country                    Area  \
0   2013.0    Unprovoked         PHILIPPINES                 Palawan   
1   2013.0    Unprovoked         PHILIPPINES                 Palawan   
2   2006.0       Boating      ATLANTIC OCEAN  300 miles from Antigua   
3   2006.0       Boating      ATLANTIC OCEAN  300 miles from Antigua   
4   2003.0          Boat        SOUTH AFRICA   Western Cape Province   
5   2003.0          Boat        SOUTH AFRICA   Western Cape Province   
6   2000.0       Invalid                 USA                 Florida   
7   2000.0       Invalid                 USA                 Florida   
8   1994.0  Sea Disaster                 USA                 Florida   
9   1994.0  Sea Disaster                 USA                 Florida   
10  1993.0       Boating       CARIBBEAN SEA                     NaN   
11  1993.0       Boating       CARIBBEAN SEA                     NaN   
12  1968.0    Unprovoked                 USA                 Florida   
13  1968.0    Unprovoked                 USA                 Florida   
14  1963.0    Unprovoked  DOMINICAN REPUBLIC           Santo Domingo   
15  1963.0    Unprovoked  DOMINICAN REPUBLIC           Santo Domingo   
16  1961.0      Provoked               ITALY         Venice Province   
17  1961.0      Provoked               ITALY         Venice Province   
18  1960.0    Unprovoked           AUSTRALIA                Tasmania   
19  1960.0    Unprovoked           AUSTRALIA                Tasmania   
20  1959.0      Provoked           AUSTRALIA                Victoria   
21  1959.0      Provoked           AUSTRALIA                Victoria   
22  1959.0    Unprovoked          MOZAMBIQUE         Maputo Province   
23  1959.0    Unprovoked          MOZAMBIQUE         Maputo Province   
24  1959.0       Invalid               ITALY                 Tuscany   
25  1959.0       Invalid               ITALY                 Tuscany   
26  1959.0          Boat           AUSTRALIA         New South Wales   
27  1959.0          Boat           AUSTRALIA         New South Wales   
28  1947.0       Boating           AUSTRALIA         New South Wales   
29  1947.0       Boating           AUSTRALIA         New South Wales   
30  1932.0      Provoked              MEXICO         Baja California   
31  1932.0      Provoked              MEXICO         Baja California   
32  2006.0       Boating      ATLANTIC OCEAN     800 miles from land   
33  2006.0       Boating      ATLANTIC OCEAN     800 miles from land   
34  2006.0       Boating      ATLANTIC OCEAN     800 miles from land   
35  1960.0       Invalid             BERMUDA                33N, 68W   
36  1960.0       Invalid             BERMUDA                33N, 68W   
37  1960.0       Invalid             BERMUDA                33N, 68W   
38  1960.0       Invalid             BERMUDA                33N, 68W   
39  1960.0       Invalid                 USA              California   
40  1960.0       Invalid                 USA              California   
41  1960.0       Invalid                 USA              California   
42  1960.0       Invalid                 USA              California   
43  1960.0       Invalid                 USA              California   

                                             Activity Sex   Age  \
0   Swimming to shore with floatioon devices after...    M   28   
1   Swimming to shore with floatioon devices after...    M   26   
2      Competing in the Woodvale Atlantic Rowing Race    M   46   
3      Competing in the Woodvale Atlantic Rowing Race    M   34   
4                                            Fishing   NaN   36   
5                                            Fishing   NaN   26   
6               Attempting to illegally enter the USA    M   23   
7               Attempting to illegally enter the USA    M   20   
8                              Adrift on refugee raft    M    7   
9                              Adrift on refugee raft    M   31   
10                          

limit_output extension: Maximum message size of 10000 exceeded with 15027 characters

In [62]:
# Concatenando o DataFrame age_final com o atck
atck = pd.concat([atck, age_final])

In [63]:
# Reordenando por ano (decrescente) e idade (crescente)
atck = atck.sort_values(['Year', 'Age'], ascending = [False, True])
atck = atck.reset_index(drop = True)
atck

Year        Type    Country               Area         Activity Sex   \
0      2018.0  Unprovoked        USA     South Carolina  Boogie boarding    M   
1      2018.0  Unprovoked        USA     South Carolina         Swimming    M   
2      2018.0  Unprovoked  AUSTRALIA  Western Australia           Wading    M   
3      2018.0  Unprovoked        USA            Georgia         Standing    F   
4      2018.0  Unprovoked        USA            Florida         Standing    M   
...       ...         ...        ...                ...              ...  ...   
25742     NaN         NaN        NaN                NaN              NaN  NaN   
25743     NaN         NaN        NaN                NaN              NaN  NaN   
25744     NaN         NaN        NaN                NaN              NaN  NaN   
25745     NaN         NaN        NaN                NaN              NaN  NaN   
25746     NaN         NaN        NaN                NaN              NaN  NaN   

       Age                                Injury Fatal (Y/N)         Species   
0       10  Injuries to lower right leg and foot           N              NaN  
1       10          Severe bite to right forearm           N              NaN  
2       10                Minor injuries to legs           N  Wobbegong shark  
3       11            Minor injury to left thigh           N              NaN  
4       12                  Minor injury to foot           N              NaN  
...    ...                                   ...         ...              ...  
25742  NaN                                   NaN         NaN              NaN  
25743  NaN                                   NaN         NaN              NaN  
25744  NaN                                   NaN         NaN              NaN  
25745  NaN                                   NaN         NaN              NaN  
25746  NaN                                   NaN         NaN              NaN  

[25747 rows x 10 columns]

### Buscando e Corrigindo outras linhas da coluna 'Age' que apresentam str

In [64]:
# Identificar quais as outras STR dentro da coluna Age para corrigir
atck['Age'].unique()

# PROBLEMAS:
# s no final do int. ex: 40s
# nan
# ?
# months
# �
# teen, Teens, mid, to, Elderly, Ca., adult, >, F, young, Both 
# MAKE LINE GREEN
# A.M

array(['10', '11', '12', '14', '15', '18', '19', '20', '21', '25', '28',
       '29', '30', '32', '33', '34', '35', '37', '38', '40s', '41', '45',
       '46', '48', '52', '54', '55', '56', '60', '69', nan, '13', '16',
       '17', '20s', '22', '24', '26', '3', '31', '39', '40', '42', '47',
       '49', '50', '51', '57', '58', '68', '7', '73', '8', '82', 'Teen',
       '23', '27', '36', '43', '59', '6', '60s', '61', '64', '65', '66',
       '72', '9', '18 months', '44', '62', '63', '70', '71', '30s', '50s',
       '53', 'teen', 26, 28, '74', '77', '5', '86', 23, 30, 34, 46,
       'Teens', 36, '8 ', '84', ' ', '� ', '21 & ?', '6�', '75',
       'mid-30s', 20, ' 30', 7, 31, ' 28', '20?', 32, "60's", '16 to 18',
       '67', '87', 'Elderly', 'Ca. 33', 'mid-20s', '45 ', '74 ', '20 ',
       '>50', '18 to 22', 'adult', 9, 12, '? & 19', '9 months',
       '25 to 35', '(adult)', '1', 33, 37, 21, 24, 27, 35, 67, 17, 19, 50,
       '"middle-age"', 'X', '2 to 3 months', 'MAKE LINE GREEN', ' 43'

In [65]:
atck.loc[atck['Age'] == "16 to 18"]

Year     Type    Country             Area Activity Sex        Age  \
2769  1987.0  Invalid  AUSTRALIA  New South Wales      NaN    F  16 to 18   

                                                 Injury Fatal (Y/N)  \
2769  Remains recovered from 4 m, 420-kg, caught 13-...         NaN   

                                               Species   
2769  Shark involvement prior to death was not confi...

In [66]:
# Fazer um Regex para identificar os elementos 100% str no AGE
padrão = '[A-z ]+'
lista_str = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_str.append(i)
    except:
        None
        
print(lista_str)

['Teen', 'Teen', 'Teen', 'teen', 'teen', 'teen', 'Teen', 'Teen', 'teen', 'Teens', 'teen', 'Teen', ' ', 'Elderly', ' ', 'adult', 'Teen', 'X', 'MAKE LINE GREEN', 'F', 'young', 'Teen', '  ', 'Teen', 'young']


In [67]:
atck.loc[atck['Age'] == 'young']

Year        Type    Country             Area  \
4891  1936.0    Provoked  AUSTRALIA  New South Wales   
6165  1779.0  Unprovoked        USA           Hawaii   

                                           Activity Sex     Age  \
4891  Touching the mouth of a supposedly dead shark    F  young   
6165                                        Surfing    M  young   

                               Injury Fatal (Y/N) Species   
4891  Finger bitten PROVOKED INCIDENT           N      NaN  
6165       FATAL, buttock lacerated             Y      NaN

In [68]:
# listas com a indexação dos elementos 100% str
lista_index_teen = ([141, 287, 540, 1028, 1381, 1683, 3634, 5030, 5567, 541, 542, 683,
                    1382, 1592, 1591, 5203, 2769])
lista_index_adult = [3441, 1959, 4059, 2989, 3785, 3620]
lista_index_young = [4891, 6165, 4444]
lista_index_elderly = [2898, 1956, 2498]


# Alterando os valores de acordo com seguintes valores estabelecidos:
 # - teen = 16 anos; elderly = 65 anos; young = 9  anos; adult = 35 anos
# Utilizando esses valores não jogo um dado fora ao mesmo tempo que não vou alterar de forma
# significativa as análises estatísticas da amostra

for idade in lista_index_teen:
    atck['Age'][idade] = 16
    
for idade in lista_index_adult:
    atck['Age'][idade] = 35

for idade in lista_index_young:
    atck['Age'][idade] = 9

for idade in lista_index_elderly:
    atck['Age'][idade] = 65

<ipython-input-68-3b61ee802b2c>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][idade] = 16
<ipython-input-68-3b61ee802b2c>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][idade] = 35
<ipython-input-68-3b61ee802b2c>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][idade] = 9
<ipython-input-68-3b61ee802b2c>:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataF

In [69]:
# Regex para identificar números + STR
padrão = '[0-9]+[ A-z]+'
lista_str1 = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_str1.append(i)
    except:
        None
print(lista_str1)

['40s', '20s', '40s', '60s', '18 months', '20s', '30s', '40s', '50s', '30s', '20s', '20s', '30s', '30s', '30s', '8 ', '30s', '30s', '20s', '20s', '45 ', '74 ', '20 ', '20s', '9 months', '7 ', '8 ', '9 ', '50s', '7 ']


In [70]:
# Definindo a indexação desses números com str
index_40s = [27, 244, 513]
index_20s = [77, 482, 849, 1442, 2884, 3059, 3298, 1922, 2990, 3391, 3531, 6022]
index_30s = [498, 762, 1463, 1560, 1654, 2633, 2634]
index_50s = [527, 5090, 3374]
# index com 60s = 274


In [71]:
# Fazendo a substituição do STR por INT na devida indexação
for a in index_40s:
    atck['Age'][a] = 40

for b in index_20s:
    atck['Age'][b] = 20

for c in index_30s:
    atck['Age'][c] = 30
    
for d in index_50s:
    atck['Age'][d] = 50

# 60s
atck['Age'][274] = 60

<ipython-input-71-f024969910af>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][a] = 40
<ipython-input-71-f024969910af>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][b] = 20
<ipython-input-71-f024969910af>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][c] = 30
<ipython-input-71-f024969910af>:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the 

In [72]:
# index 343 = 18 months = 1.5 anos
# index 3590 = 9 months = 0.75 anos
atck['Age'][343] = 1.5
atck['Age'][3590] = 0.75

<ipython-input-72-d1dfb3d6e203>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][343] = 1.5
<ipython-input-72-d1dfb3d6e203>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Age'][3590] = 0.75


In [73]:
# Regex para identificar o padrão STR + números
padrão = '[A-z ]+[0-9]+'
lista_str2 = list()
for i in atck['Age']:
    try:
        if i in re.findall(padrão, i):
            lista_str2.append(i)
    except:
        None
print(lista_str2)

[' 30', ' 28', ' 43', 'Both 11']


In [74]:
atck.loc[atck['Age'] == 'Both 11']
# index 4806 = Both 11 - Dois casos com a mesma idade em uma única linha

# Criando um DataFrame com as informações dessa linha e corrigindo o Age para 11
new1 = pd.DataFrame(info_mult(4810))
new1 = new1.transpose()
new1.columns = atck.columns
new1['Age'][0] = 11

# Subindo 2x as informações para o dataframe principal e dropando a informação duplicada
atck = pd.concat([atck, new1])
atck = pd.concat([atck, new1])
atck = atck.drop(4810)

atck

Year        Type    Country               Area         Activity Sex   \
0      2018.0  Unprovoked        USA     South Carolina  Boogie boarding    M   
1      2018.0  Unprovoked        USA     South Carolina         Swimming    M   
2      2018.0  Unprovoked  AUSTRALIA  Western Australia           Wading    M   
3      2018.0  Unprovoked        USA            Georgia         Standing    F   
4      2018.0  Unprovoked        USA            Florida         Standing    M   
...       ...         ...        ...                ...              ...  ...   
25744     NaN         NaN        NaN                NaN              NaN  NaN   
25745     NaN         NaN        NaN                NaN              NaN  NaN   
25746     NaN         NaN        NaN                NaN              NaN  NaN   
0      1939.0  Unprovoked  AUSTRALIA    New South Wales  Scooping prawns    M   
0      1939.0  Unprovoked  AUSTRALIA    New South Wales  Scooping prawns    M   

       Age                                Injury Fatal (Y/N)         Species   
0       10  Injuries to lower right leg and foot           N              NaN  
1       10          Severe bite to right forearm           N              NaN  
2       10                Minor injuries to legs           N  Wobbegong shark  
3       11            Minor injury to left thigh           N              NaN  
4       12                  Minor injury to foot           N              NaN  
...    ...                                   ...         ...              ...  
25744  NaN                                   NaN         NaN              NaN  
25745  NaN                                   NaN         NaN              NaN  
25746  NaN                                   NaN         NaN              NaN  
0       11                             No injury           N         7' shark  
0       11                             No injury           N         7' shark  

[25748 rows x 10 columns]

In [75]:
#Reordenando por ano (decrescente) e idade (crescente)
atck = atck.sort_values(['Year', 'Age'], ascending = [False, True])
atck = atck.reset_index(drop = True)
atck

Year        Type        Country               Area         Activity  \
0      2018.0  Unprovoked  NEW CALEDONIA                NaN     Spearfishing   
1      2018.0  Unprovoked            USA     South Carolina  Boogie boarding   
2      2018.0  Unprovoked            USA     South Carolina         Swimming   
3      2018.0  Unprovoked      AUSTRALIA  Western Australia           Wading   
4      2018.0  Unprovoked            USA            Georgia         Standing   
...       ...         ...            ...                ...              ...   
25743     NaN         NaN            NaN                NaN              NaN   
25744     NaN         NaN            NaN                NaN              NaN   
25745     NaN         NaN            NaN                NaN              NaN   
25746     NaN         NaN            NaN                NaN              NaN   
25747     NaN         NaN            NaN                NaN              NaN   

      Sex   Age                                Injury Fatal (Y/N)  \
0        M   40           Injuries to abdomen and arm           N   
1        M   10  Injuries to lower right leg and foot           N   
2        M   10          Severe bite to right forearm           N   
3        M   10                Minor injuries to legs           N   
4        F   11            Minor injury to left thigh           N   
...    ...  ...                                   ...         ...   
25743  NaN  NaN                                   NaN         NaN   
25744  NaN  NaN                                   NaN         NaN   
25745  NaN  NaN                                   NaN         NaN   
25746  NaN  NaN                                   NaN         NaN   
25747  NaN  NaN                                   NaN         NaN   

               Species   
0      Bull shark, 3 m   
1                   NaN  
2                   NaN  
3       Wobbegong shark  
4                   NaN  
...                 ...  
25743               NaN  
25744               NaN  
25745               NaN  
25746               NaN  
25747               NaN  

[25748 rows x 10 columns]

### Transformando a coluna 'Age em float

In [76]:
import numpy as np
atck['Age'].unique()

array([40, '10', '11', '12', '14', '15', '18', '19', '20', '21', '25',
       '28', '29', '30', '32', '33', '34', '35', '37', '38', '41', '45',
       '46', '48', '52', '54', '55', '56', '60', '69', nan, 16, 20, '13',
       '16', '17', '22', '24', '26', '3', '31', '39', '40', '42', '47',
       '49', '50', '51', '57', '58', '68', '7', '73', '8', '82', 60, '23',
       '27', '36', '43', '59', '6', '61', '64', '65', '66', '72', '9',
       1.5, '44', '62', '63', '70', '71', 30, 50, '53', 26, 28, '74',
       '77', '5', '86', 23, 34, 46, 36, '8 ', '84', ' ', '� ', 35, 65,
       '75', ' 30', 7, 31, ' 28', '20?', 32, '67', '87', '45 ', '74 ',
       '20 ', 9, 12, 0.75, '1', 33, 37, 21, 24, 27, 67, 17, 19, 'X',
       '2 to 3 months', 'MAKE LINE GREEN', ' 43', '81', '7 ', '78', 'F',
       11, '9 ', '  ', 'A.M.'], dtype=object)

In [77]:
def age_float(value):
    try:
        return int(value)
    except ValueError as e:
        return np.nan
    
atck['Age'] = atck['Age'].map(age_float)

In [78]:
atck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25748 entries, 0 to 25747
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6325 non-null   object 
 1   Type         6323 non-null   object 
 2   Country      6277 non-null   object 
 3   Area         5871 non-null   object 
 4   Activity     5783 non-null   object 
 5   Sex          5758 non-null   object 
 6   Age          3486 non-null   float64
 7   Injury       6299 non-null   object 
 8   Fatal (Y/N)  5779 non-null   object 
 9   Species      3484 non-null   object 
dtypes: float64(1), object(9)
memory usage: 2.0+ MB


## Transformando as colunas Year em float

In [79]:
atck['Year'] = atck['Year'].astype(float, errors = 'raise')

atck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25748 entries, 0 to 25747
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6325 non-null   float64
 1   Type         6323 non-null   object 
 2   Country      6277 non-null   object 
 3   Area         5871 non-null   object 
 4   Activity     5783 non-null   object 
 5   Sex          5758 non-null   object 
 6   Age          3486 non-null   float64
 7   Injury       6299 non-null   object 
 8   Fatal (Y/N)  5779 non-null   object 
 9   Species      3484 non-null   object 
dtypes: float64(2), object(8)
memory usage: 2.0+ MB


## Analisando as outras colunas

In [80]:
atck.columns

Index(['Year', 'Type', 'Country', 'Area', 'Activity', 'Sex ', 'Age', 'Injury',
       'Fatal (Y/N)', 'Species '],
      dtype='object')

In [81]:
# Dados parecem estar ok para serem avaliados
print(atck['Type'].value_counts())
print(atck['Type'].value_counts(normalize=True).map("{:.1%}".format))

Unprovoked      4601
Provoked         577
Invalid          556
Sea Disaster     240
Boating          207
Boat             139
Questionable       2
Boatomg            1
Name: Type, dtype: int64
Unprovoked      72.8%
Provoked         9.1%
Invalid          8.8%
Sea Disaster     3.8%
Boating          3.3%
Boat             2.2%
Questionable     0.0%
Boatomg          0.0%
Name: Type, dtype: object


In [82]:
print(atck['Country'].value_counts())
print(atck['Country'].value_counts(normalize=True).map("{:.1%}".format))

USA                             2235
AUSTRALIA                       1343
SOUTH AFRICA                     580
PAPUA NEW GUINEA                 134
NEW ZEALAND                      128
                                ... 
ALGERIA                            1
EQUATORIAL GUINEA / CAMEROON       1
BANGLADESH                         1
 PHILIPPINES                       1
GHANA                              1
Name: Country, Length: 212, dtype: int64
USA                             35.6%
AUSTRALIA                       21.4%
SOUTH AFRICA                     9.2%
PAPUA NEW GUINEA                 2.1%
NEW ZEALAND                      2.0%
                                ...  
ALGERIA                          0.0%
EQUATORIAL GUINEA / CAMEROON     0.0%
BANGLADESH                       0.0%
 PHILIPPINES                     0.0%
GHANA                            0.0%
Name: Country, Length: 212, dtype: object


In [83]:
print(atck['Area'].value_counts())
print(atck['Area'].value_counts(normalize=True).map("{:.1%}".format))

Florida             1040
New South Wales      489
Queensland           311
Hawaii               298
California           293
                    ... 
 La Libertad           1
Romblon Province       1
Savona                 1
Bikini Atoll           1
Genoa  Province        1
Name: Area, Length: 825, dtype: int64
Florida             17.7%
New South Wales      8.3%
Queensland           5.3%
Hawaii               5.1%
California           5.0%
                    ...  
 La Libertad         0.0%
Romblon Province     0.0%
Savona               0.0%
Bikini Atoll         0.0%
Genoa  Province      0.0%
Name: Area, Length: 825, dtype: object


In [84]:
print(atck['Activity'].value_counts())
print(atck['Activity'].value_counts(normalize=True).map("{:.1%}".format))

Surfing                                                                   971
Swimming                                                                  869
Fishing                                                                   434
Spearfishing                                                              333
Bathing                                                                   162
                                                                         ... 
Swimming near jetty with 2' piece of wood                                   1
63' fishing boat Sno' Bay foundered                                         1
American minesweeper USS YMS-350 lost in a typhoon - swimming to shore      1
Adrift on life raft                                                         1
Lobstering                                                                  1
Name: Activity, Length: 1532, dtype: int64
Surfing                                                                   16.8%
Swimming           

In [85]:
atck['Sex '].value_counts()
# Vou melhorar isso aqui

M      5114
F       638
M         2
N         2
lli       1
.         1
Name: Sex , dtype: int64

In [86]:
print(atck['Age'].value_counts())
print(atck['Age'].value_counts(normalize=True).map("{:.1%}".format))

16.0    156
17.0    156
20.0    155
18.0    151
19.0    143
       ... 
86.0      1
72.0      1
87.0      1
84.0      1
0.0       1
Name: Age, Length: 81, dtype: int64
16.0    4.5%
17.0    4.5%
20.0    4.4%
18.0    4.3%
19.0    4.1%
        ... 
86.0    0.0%
72.0    0.0%
87.0    0.0%
84.0    0.0%
0.0     0.0%
Name: Age, Length: 81, dtype: object


In [87]:
atck['Injury'].value_counts()
# Vou tirar

FATAL                                                                                                                                                                                                                                      803
Survived                                                                                                                                                                                                                                    97
Foot bitten                                                                                                                                                                                                                                 87
No injury                                                                                                                                                                                                                                   83
Leg bitten                                  

In [88]:
atck['Fatal (Y/N)'].value_counts()
# Vou melhorar isso aqui

N          4305
Y          1392
UNKNOWN      71
 N            7
y             1
M             1
N             1
2017          1
Name: Fatal (Y/N), dtype: int64

In [89]:
atck['Species '].value_counts()
# Vou tirar

White shark                                           162
Shark involvement prior to death was not confirmed    111
Invalid                                               102
Shark involvement not confirmed                        88
Tiger shark                                            73
                                                     ... 
Said to involve 2 sharks: 5.2 m & 6 m [17' & 20']       1
Grey reef shark, 2 m [6.75']                            1
2.3 m [7.5'] white shark, identified by M. Smale        1
2.1 m [7'], 140-lb reef shark                           1
2.1 to 2.4 m [7' to 8'] shark                           1
Name: Species , Length: 1549, dtype: int64

### Conclusão da Análise das Colunas

In [90]:
# As colunas 'Injury' e 'Species ' não apresentam um padrão de preenchimento
# informações se tornam incompletas para uma análise. Dropar.

atck = atck.drop(columns=['Species '])
atck = atck.drop(columns=['Injury'])

atck

Year        Type        Country               Area         Activity  \
0      2018.0  Unprovoked  NEW CALEDONIA                NaN     Spearfishing   
1      2018.0  Unprovoked            USA     South Carolina  Boogie boarding   
2      2018.0  Unprovoked            USA     South Carolina         Swimming   
3      2018.0  Unprovoked      AUSTRALIA  Western Australia           Wading   
4      2018.0  Unprovoked            USA            Georgia         Standing   
...       ...         ...            ...                ...              ...   
25743     NaN         NaN            NaN                NaN              NaN   
25744     NaN         NaN            NaN                NaN              NaN   
25745     NaN         NaN            NaN                NaN              NaN   
25746     NaN         NaN            NaN                NaN              NaN   
25747     NaN         NaN            NaN                NaN              NaN   

      Sex    Age Fatal (Y/N)  
0        M  40.0           N  
1        M  10.0           N  
2        M  10.0           N  
3        M  10.0           N  
4        F  11.0           N  
...    ...   ...         ...  
25743  NaN   NaN         NaN  
25744  NaN   NaN         NaN  
25745  NaN   NaN         NaN  
25746  NaN   NaN         NaN  
25747  NaN   NaN         NaN  

[25748 rows x 8 columns]

In [91]:
# Pequenas correções na coluna 'Fatal (Y/N)'.
# Identificar a indexação e corrigir:

atck.loc[atck['Fatal (Y/N)'] == ' N']
lista_fatal_N = [3967, 3974, 3983, 3985, 3987, 3993, 4006, 219, 5450]

for i in lista_fatal_N:
    atck['Fatal (Y/N)'][i] = 'N'

atck['Fatal (Y/N)'][799] = np.nan
atck['Fatal (Y/N)'][6141] = 'Y'

print(atck['Fatal (Y/N)'].value_counts())
print(atck['Fatal (Y/N)'].value_counts(normalize=True).map("{:.1%}".format))

N          4314
Y          1393
UNKNOWN      71
Name: Fatal (Y/N), dtype: int64
N          74.7%
Y          24.1%
UNKNOWN     1.2%
Name: Fatal (Y/N), dtype: object


<ipython-input-91-386ad7350e5d>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Fatal (Y/N)'][i] = 'N'
<ipython-input-91-386ad7350e5d>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Fatal (Y/N)'][799] = np.nan
<ipython-input-91-386ad7350e5d>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Fatal (Y/N)'][6141] = 'Y'


In [92]:
# A coluna sex também tem pequenas correções
# Fazer a mesma coisa que fiz no Fatal Y/N

atck['Sex '][6154] = 'M'
atck['Sex '][4967] = 'M'
atck['Sex '][519] = 'M'
atck['Sex '][1605] = 'M'
atck['Sex '][5460] = np.nan
atck['Sex '][1660] = np.nan

print(atck['Sex '].value_counts())
print(atck['Sex '].value_counts(normalize=True).map("{:.1%}".format))

M    5118
F     638
Name: Sex , dtype: int64
M    88.9%
F    11.1%
Name: Sex , dtype: object


<ipython-input-92-a979153abda5>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Sex '][6154] = 'M'
<ipython-input-92-a979153abda5>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Sex '][4967] = 'M'
<ipython-input-92-a979153abda5>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  atck['Sex '][519] = 'M'
<ipython-input-92-a979153abda5>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFr

## Limpando as linhas sem nenhuma informação

In [93]:
atck.isna()

Year   Type  Country   Area  Activity   Sex     Age  Fatal (Y/N)
0      False  False    False   True     False  False  False        False
1      False  False    False  False     False  False  False        False
2      False  False    False  False     False  False  False        False
3      False  False    False  False     False  False  False        False
4      False  False    False  False     False  False  False        False
...      ...    ...      ...    ...       ...    ...    ...          ...
25743   True   True     True   True      True   True   True         True
25744   True   True     True   True      True   True   True         True
25745   True   True     True   True      True   True   True         True
25746   True   True     True   True      True   True   True         True
25747   True   True     True   True      True   True   True         True

[25748 rows x 8 columns]

In [94]:
atck['info_nulas'] = atck.isna().sum(axis = 1)

In [95]:
atck['info_nulas'].describe()

count    25748.000000
mean         6.229027
std          3.137063
min          0.000000
25%          8.000000
50%          8.000000
75%          8.000000
max          8.000000
Name: info_nulas, dtype: float64

In [96]:
atck['index'] = atck.index

In [97]:
sem_info = atck[atck['info_nulas'] == 8]
sem_info

Year Type Country Area Activity Sex   Age Fatal (Y/N)  info_nulas  \
6327    NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
6328    NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
6329    NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
6330    NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
6331    NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
...     ...  ...     ...  ...      ...  ...  ...         ...         ...   
25743   NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
25744   NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
25745   NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
25746   NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   
25747   NaN  NaN     NaN  NaN      NaN  NaN  NaN         NaN           8   

       index  
6327    6327  
6328    6328  
6329    6329  
6330    6330  
6331    6331  
...      ...  
25743  25743  
25744  25744  
25745  25745  
25746  25746  
25747  25747  

[19421 rows x 10 columns]

In [98]:
# O arquivo apresenta 19421 linhas sem nenhuma informação
# Dropar tudo!
dropar = list()
for i in range(len(sem_info)):
    dropar.append(sem_info['index'].iloc[i])


In [99]:
# Conferindo... ok
print(dropar[0])
print(dropar[-1])
print(len(dropar))

6327
25747
19421


In [100]:
#Essa célula demora mais de 2min para rodar... liberar quando estiver todo o resto pronto
for i in dropar:
    atck = atck.drop(i)

In [101]:
atck

Year        Type        Country               Area         Activity  \
0     2018.0  Unprovoked  NEW CALEDONIA                NaN     Spearfishing   
1     2018.0  Unprovoked            USA     South Carolina  Boogie boarding   
2     2018.0  Unprovoked            USA     South Carolina         Swimming   
3     2018.0  Unprovoked      AUSTRALIA  Western Australia           Wading   
4     2018.0  Unprovoked            USA            Georgia         Standing   
...      ...         ...            ...                ...              ...   
6322     0.0  Unprovoked      AUSTRALIA  Western Australia     Pearl diving   
6323     0.0  Unprovoked            USA     North Carolina         Swimming   
6324     0.0  Unprovoked         PANAMA                NaN              NaN   
6325     NaN     Invalid      AUSTRALIA         Queensland     Spearfishing   
6326     NaN  Unprovoked        ENGLAND         Cumberland         Swimming   

     Sex    Age Fatal (Y/N)  info_nulas  index  
0       M  40.0           N           1      0  
1       M  10.0           N           0      1  
2       M  10.0           N           0      2  
3       M  10.0           N           0      3  
4       F  11.0           N           0      4  
...   ...   ...         ...         ...    ...  
6322    M   NaN           Y           1   6322  
6323    M   NaN           Y           1   6323  
6324    M   NaN           Y           3   6324  
6325    M  35.0         NaN           2   6325  
6326    M   NaN           Y           2   6326  

[6327 rows x 10 columns]

In [102]:
# dropando a coluna info_nulas e index

atck = atck.drop(columns=['info_nulas'])
atck = atck.drop(columns=['index'])

atck

Year        Type        Country               Area         Activity  \
0     2018.0  Unprovoked  NEW CALEDONIA                NaN     Spearfishing   
1     2018.0  Unprovoked            USA     South Carolina  Boogie boarding   
2     2018.0  Unprovoked            USA     South Carolina         Swimming   
3     2018.0  Unprovoked      AUSTRALIA  Western Australia           Wading   
4     2018.0  Unprovoked            USA            Georgia         Standing   
...      ...         ...            ...                ...              ...   
6322     0.0  Unprovoked      AUSTRALIA  Western Australia     Pearl diving   
6323     0.0  Unprovoked            USA     North Carolina         Swimming   
6324     0.0  Unprovoked         PANAMA                NaN              NaN   
6325     NaN     Invalid      AUSTRALIA         Queensland     Spearfishing   
6326     NaN  Unprovoked        ENGLAND         Cumberland         Swimming   

     Sex    Age Fatal (Y/N)  
0       M  40.0           N  
1       M  10.0           N  
2       M  10.0           N  
3       M  10.0           N  
4       F  11.0           N  
...   ...   ...         ...  
6322    M   NaN           Y  
6323    M   NaN           Y  
6324    M   NaN           Y  
6325    M  35.0         NaN  
6326    M   NaN           Y  

[6327 rows x 8 columns]

# Conclusões
- O arquivo foi aberto com 25723 linhas e 24 colunas
- Foi encerrado com 6327 linhas e 8 colunas
- Abaixo os principais dados descritivos retirados dessa análise

In [103]:
print(atck['Year'].value_counts())
print(atck['Year'].value_counts(normalize=True).map("{:.1%}".format).head(30))
print('')
print(atck['Type'].value_counts())
print(atck['Type'].value_counts(normalize=True).map("{:.1%}".format))
print('')
print(atck['Country'].value_counts())
print(atck['Country'].value_counts(normalize=True).map("{:.1%}".format))
print('')
print(atck['Area'].value_counts())
print(atck['Area'].value_counts(normalize=True).map("{:.1%}".format))
print('')
print(atck['Activity'].value_counts())
print(atck['Activity'].value_counts(normalize=True).map("{:.1%}".format))
print('')
print(atck['Age'].value_counts())
print(atck['Age'].value_counts(normalize=True).map("{:.1%}".format).head(15))
print('')
print(atck['Fatal (Y/N)'].value_counts())
print(atck['Fatal (Y/N)'].value_counts(normalize=True).map("{:.1%}".format))
print('')
print(atck['Sex '].value_counts())
print(atck['Sex '].value_counts(normalize=True).map("{:.1%}".format))

2015.0    143
2017.0    136
2016.0    130
2011.0    128
2014.0    127
         ... 
1767.0      1
1816.0      1
1812.0      1
1822.0      1
1748.0      1
Name: Year, Length: 249, dtype: int64
2015.0    2.3%
2017.0    2.2%
2016.0    2.1%
2011.0    2.0%
2014.0    2.0%
0.0       2.0%
2013.0    1.9%
2008.0    1.9%
2009.0    1.9%
2012.0    1.8%
2007.0    1.8%
2006.0    1.7%
2005.0    1.6%
1960.0    1.6%
2010.0    1.6%
2000.0    1.5%
1959.0    1.5%
2003.0    1.5%
2001.0    1.5%
2004.0    1.5%
2002.0    1.4%
1962.0    1.4%
1961.0    1.2%
1995.0    1.2%
1964.0    1.0%
1999.0    1.0%
1998.0    1.0%
1963.0    1.0%
1996.0    1.0%
1966.0    0.9%
Name: Year, dtype: object

Unprovoked      4601
Provoked         577
Invalid          556
Sea Disaster     240
Boating          207
Boat             139
Questionable       2
Boatomg            1
Name: Type, dtype: int64
Unprovoked      72.8%
Provoked         9.1%
Invalid          8.8%
Sea Disaster     3.8%
Boating          3.3%
Boat             2.2%
Questi

In [104]:
print(atck_full['Location'].value_counts().head(20))
print(atck_full['Location'].value_counts(normalize=True).map("{:.1%}".format))

New Smyrna Beach, Volusia County                 163
Daytona Beach, Volusia County                     30
Ponce Inlet, Volusia County                       20
Melbourne Beach, Brevard County                   18
Myrtle Beach, Horry County                        17
Durban                                            16
Boa Viagem, Recife                                14
Isle of Palms, Charleston County                  14
Cocoa Beach, Brevard County                       12
Ponce Inlet, New Smyrna Beach, Volusia County     12
Piedade                                           11
Ahvaz, on the Karun River                         11
Cocoa Beach, Brevard  County                      11
Ormond Beach, Volusia County                      11
Palm Beach, Palm Beach County                     11
Mossel Bay                                        11
Sydney                                            10
Xai Xai                                           10
Sydney Harbor                                 

# Análise dos Últimos 10 anos
- Os dados totais tem datas extremamente antigas.
- Se aproximarmos para os anos de 1997 até 2018, veremos diferenças?

In [123]:
mkt = (atck['Year'], atck['Area'], 
       atck['Activity'], atck['Age'], atck['Fatal (Y/N)'])
mkt = pd.DataFrame(mkt)
mkt = mkt.transpose()

In [124]:
mkt = mkt[mkt['Year'] > 1997]

In [125]:
mkt

Year               Area  \
0     2018.0                NaN   
1     2018.0     South Carolina   
2     2018.0     South Carolina   
3     2018.0  Western Australia   
4     2018.0            Georgia   
...      ...                ...   
2209  1998.0     South Province   
2210  1998.0         Grand'Anse   
2211  1998.0    South Australia   
2212  1998.0          Beaufonds   
2213  1998.0           Virginia   

                                               Activity   Age Fatal (Y/N)  
0                                          Spearfishing  40.0           N  
1                                       Boogie boarding  10.0           N  
2                                              Swimming  10.0           N  
3                                                Wading  10.0           N  
4                                              Standing  11.0           N  
...                                                 ...   ...         ...  
2209                                        Windsurfing   NaN           N  
2210                                            Bathing   NaN           Y  
2211                                            Surfing   NaN           N  
2212                                             Diving   NaN           Y  
2213  Spearfishing on scuba & transferring fish onto...   NaN           N  

[2214 rows x 5 columns]

In [198]:
# 538 ataques em indivíduos entre 15 e 25 anos = 24,3%
# 38% dos ataques em individuos entre 14 e 26 anos = 627 ataques
# 385 ataques em indivíduos entre 27 e 39 anos = 23,3%

#31,7% em surfistas
# 13,8% em nadadores
# 5,9% em pesca submarina

# 27,4% na Florida
# 7,5% New South Wales
# 6,3% Hawaii
# 5,5% California
# 4,4% Western Australia

mkt['Area'].value_counts(normalize=True).map('{:.1%}'.format)

Florida                      27.4%
New South Wales               7.5%
Hawaii                        6.3%
California                    5.5%
Western Australia             4.4%
                             ...  
Marquesas                     0.0%
Santiago de Cuba Province     0.0%
Devon                         0.0%
Phuket                        0.0%
Bimini Islands                0.0%
Name: Area, Length: 286, dtype: object

## Não mudou nada!